File: src/my_py_pkg/my_py_pkg/battery_node.py
#!/usr/bin/env python3

import time
import rclpy
from rclpy.node import Node

from functools import partial

from my_robot_interfaces.srv import SetLed


class BatteryNode(Node):

    def __init__(self):
        super().__init__("battery")
        self.charge_ = 6.0
        self.recharge_ = False
        self.timer_ = self.create_timer(1.0, self.update_battery) 

    def update_battery(self):
        if(self.recharge_):
            self.charge_ += 1.0
        else:
            self.charge_ -= 1.5
        
        self.get_logger().info("Battery status: {}".format(self.charge_))

        if(self.charge_ == 6.0):
            self.recharge_ = False # charge complete
            self.get_logger().info("Battery fully charge: request for turning off led")
            self.call_set_led_server(2, self.recharge_)

        elif(self.charge_ == 0.0):
            self.recharge_ = True # recharge
            self.get_logger().info("Battery recharging: request for turning on led")
            self.call_set_led_server(2, self.recharge_)

    def call_set_led_server(self, led_num, led_state):
        client = self.create_client(SetLed, "set_led")

        while(not client.wait_for_service(1.0)):
            self.get_logger().warn("Waiting for server set_led to be up")
        
        '''Service is up now'''
        request = SetLed.Request()
        request.led_num = led_num
        request.state = led_state

        future = client.call_async(request)
        future.add_done_callback(partial(self.callback_set_led, led_num=led_num, led_state=led_state))

    def callback_set_led(self, future, led_num, led_state):
        try:
            response = future.result()
            self.get_logger().info(
                "Response is {0} for request \"led_num: {1}, state: {2}\"".format(response.success, led_num, led_state))

        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))


def main(args=None):
    
    rclpy.init(args=args)  # first line to be written in any ROS2 .py node
    node = BatteryNode()
    rclpy.spin(node)  # allow node to continue to be alive
    rclpy.shutdown()  # last line of any ROS2 .py node

if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/battery_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/srv/set_led.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

using std::vector;
using std::chrono::seconds;
using std::thread;
using std::bind;
using std::string;

using my_robot_interfaces::srv::SetLed;

class BatteryNode : public rclcpp::Node
{
public:
    BatteryNode() : Node("battery"), charge_(6.0), recharge_(false)
    {
        timer_ = this->create_wall_timer(seconds(1),
                                         bind(&BatteryNode::chargeCycle, this));
    }

private:
    void chargeCycle()
    {
        charge_ += (recharge_)? 1 : -1.5;

        RCLCPP_INFO(this->get_logger(), "Battery status %.1f", charge_);
        
        if(charge_ == 6.0){
            recharge_ = false;
            RCLCPP_INFO(this->get_logger(), ("Battery fully charge: request for turning off led"));
            threads.push_back(thread(bind(&BatteryNode::callSetLedServer, this, 2, false)));
        }else if(charge_ == 0.0){
            recharge_ = true;
            RCLCPP_INFO(this->get_logger(), ("Battery recharging: request for turning on led"));
            threads.push_back(thread(bind(&BatteryNode::callSetLedServer, this, 2, true)));
        }
            
    }

    void callSetLedServer(int led_num, bool led_state)
    {
        rclcpp::Client<SetLed>::SharedPtr client_ = this->create_client<SetLed>("set_led");
        while(!client_->wait_for_service(std::chrono::seconds(1)))
            RCLCPP_WARN(this->get_logger(), "Waiting for Set Led server to be up");

        auto request = std::make_shared<SetLed::Request>();
        request->led_num = led_num;
        request->state = led_state;

        auto future = client_->async_send_request(request);

        string state_s = (request->state)? "true" : "false";
        try{
            auto response = future.get();
            string success_s = (response->success)? "true" : "false";
            RCLCPP_INFO(this->get_logger(), "Response is " + success_s + " for led_num: %d, state: " + state_s, request->led_num);

        }catch(const std::exception &e){
            RCLCPP_ERROR(this->get_logger(), "Error while calling server for led_num: %d, state: " +state_s, request->led_num);
        }
    }

    float charge_;
    bool recharge_;
    rclcpp::TimerBase::SharedPtr timer_;
    vector<thread> threads;

};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    //shared pointer, you don't need to delete manually the instance when object is not referenced (like GC in Java)
    auto node = std::make_shared<BatteryNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}